### PR TITLE
Organize the sub-topics in data analysis and transform design.

### PR DESCRIPTION
**Pull request by brightcoder01**
The root issue for the data analysis and transform discussion is #1670

### CODE REVIEW

File: docs/designs/data_transform.md [hunk 1]
```diff
@@ -84,28 +89,13 @@ SELECT * FROM new_plates TO PREDICT price USING a_model COLUMN diameter
...
-
-## Challenge
-
-* In some ML systems, users use different transform code for training and serving. It may cause inconsistency between these two stages. Consistency between training and serving is the key point of data transform. Users write the transform code only once. And then the same logic can run in batch mode for training and in real time mode for serving.  
-* We can not execute the transform code directly to transform the data records. The statistical values are required for some transform functions such as mean and variance for standardization. They need be analyzed from the entire dataset at runtime to make the transform logic concrete.  
-* The transform logic is very flexible. How do we design a SQL extended syntax to fully express the logic elegantly?  
+From the above examples, we see that a challenge is that the `TO TRAIN` clause must be able to save the input field names and the data analysis results and transformation steps together with the model, so to make sure that the predictions using the same data transformation as the training.
```
Comment (skydoorkai) [inline docs/designs/data_transform.md:92]: One challenge is to guarantee that predictions are using the same data transformation as the training. Thus, the `TO TRAIN` clause must be able to save the input field names and the data analysis results and transformation steps together with the model.

File: docs/designs/data_transform.md [hunk 2]
```diff
@@ -84,28 +89,13 @@ SELECT * FROM new_plates TO PREDICT price USING a_model COLUMN diameter
...
+[TensorFlow Transform](https://www.tensorflow.org/tfx/transform/get_started) is the open source solution for data transform in [TensorFlow Extended](https://www.tensorflow.org/tfx/guide). Users need write a Python function `preprocess_fn` to define the preprocess logic. The preprocessing function contains two groups of API calls: TensorFlow Transform Analyzers and TensorFlow Ops. Analyzer will do the statistical work on the training dataset once and convert to constant tensors. And then the statistical value and TensorFlow Ops will make the concrete transform logic as a TensorFlow graph and then convert the data record one by one. The graph will be used for both training and serving.
```
Comment (skydoorkai) [inline docs/designs/data_transform.md:98]: need write-> need to write

File: docs/designs/data_transform.md [hunk 3]
```diff
@@ -119,49 +109,124 @@ def preprocess_fn(inputs):
 
 From users' perspective, SQLFlow users prefer to write SQL instead of python. It's not user-friendly if we integrate TF Transform with SQLFlow directly.  
 
-### SQL
+### Internal System
+
+The feature engineering library in the internal system is configuration driven. It contains some primitive transform ops and users compose the transform logic with configuration file. A part of the parameters in the configuration is the statistical value. Users need to do analysis work on the dataset manually using SQL at first and then complete the configuration. What's more, the development work of auto analysis is also on-going.
```
Comment (skydoorkai) [inline docs/designs/data_transform.md]: with configuration file.  -> with a configuration file.
statistical value  -> statistical values

File: docs/designs/data_transform.md [hunk 4]
```diff
@@ -119,49 +109,124 @@ def preprocess_fn(inputs):
...
+![transform_training_inference](../images/transform_training_inference.png)
+
+From the perspective of SQLFLow, SQL can naturally support statistical work just like the analyzer. [Feature column API](https://tensorflow.google.cn/api_docs/python/tf/feature_column) and [keras preprocessing layer](https://github.com/tensorflow/community/pull/188) can take charge of the transform work as transformer. We plan to use SQL and feature column/keras preprocessing layer together to do the data analysis and transform work.  
 
-User can write SQL to do the analysis and transform work with [built-in functions](https://www.alibabacloud.com/help/doc-detail/96342.htm?spm=a2c63.p38356.b99.111.27e27309rgC5m1) or [UDF (User Defined Function)](https://www.alibabacloud.com/blog/udf-development-guide-with-maxcompute-studio_594738). The following is the SQL expression to normalize the column `capital_gain`.  
+Since we use SQL to do the analysis work, SQL requires the table schema to be wide - one column one feature. So, we will normalize the table schema at first.
```
Comment (skydoorkai) [inline docs/designs/data_transform.md]: one column one feature -> one column per feature

File: docs/designs/data_transform.md [hunk 5]
```diff
@@ -119,49 +109,124 @@ def preprocess_fn(inputs):
...
+
+From the perspective of SQLFLow, SQL can naturally support statistical work just like the analyzer. [Feature column API](https://tensorflow.google.cn/api_docs/python/tf/feature_column) and [keras preprocessing layer](https://github.com/tensorflow/community/pull/188) can take charge of the transform work as transformer. We plan to use SQL and feature column/keras preprocessing layer together to do the data analysis and transform work.  
 
-User can write SQL to do the analysis and transform work with [built-in functions](https://www.alibabacloud.com/help/doc-detail/96342.htm?spm=a2c63.p38356.b99.111.27e27309rgC5m1) or [UDF (User Defined Function)](https://www.alibabacloud.com/blog/udf-development-guide-with-maxcompute-studio_594738). The following is the SQL expression to normalize the column `capital_gain`.  
+Since we use SQL to do the analysis work, SQL requires the table schema to be wide - one column one feature. So, we will normalize the table schema at first.  
+
+### Normalize Table Schema to Be Wide
+
+Wide table means that it only store one feature or label in one column. It's more friendly to SQL for data analysis.
```
Comment (skydoorkai) [inline docs/designs/data_transform.md]: store -> stores

File: docs/designs/data_transform.md (hunk 3 above)
Comment (brightcoder01) [reply docs/designs/data_transform.md]: Good suggestion, updated.

File: docs/designs/data_transform.md (hunk 4 above)
Comment (brightcoder01) [reply docs/designs/data_transform.md]: Good suggestion, updated.

File: docs/designs/data_transform.md (hunk 5 above)
Comment (brightcoder01) [reply docs/designs/data_transform.md]: Good suggestion, updated.

File: docs/designs/data_transform.md (hunk 2 above)
Comment (brightcoder01) [reply docs/designs/data_transform.md:98]: Thanks for comments, "need do sth" is also fine.

File: docs/designs/data_transform.md (hunk 1 above)
Comment (brightcoder01) [reply docs/designs/data_transform.md:92]: IMO, the original statement is also fine.

**Review by terrytangyuan — APPROVED** [review]

**Pull request merged**